Implement building sorting by distance with location service integra…

mmc03-ucb opened the pull request:
…tion

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingInteractor.swift:
```diff
@@ -5,18 +5,67 @@
+  private let locationService: LocationService
+
+  private func calculateDistance(from location: Location, to building: Building) -> CLLocationDistance {
+    let userLocation = CLLocation(latitude: location.latitude, longitude: location.longitude)
+    let buildingLocation = CLLocation(latitude: building.latitude, longitude: building.longitude)
+    return userLocation.distance(from: buildingLocation)
```
We shouldn't have "platform" implementation details like using CoreLocation types/methods leaking into the API of the interactor. This is to allow flexibility in changing the underlying implementation in future. Are you able to perform the distance calculation with your own return type? If not you could create a collaborator unit (think of a helper function contained within a struct for e.g.) that abstracts away this logic.

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingInteractor.swift:
```diff
@@ -5,18 +5,67 @@
 //  Created by Anh Nguyen on 27/4/2025.
 //
 
+import CoreLocation
```
See the comment regarding distance calculation, ideally we shouldn't be importing platform frameworks in our interactors.

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingInteractor.swift:
```diff
@@ -5,18 +5,67 @@
+
+  // MARK: Lifecycle
+
+  init(buildingService: BuildingService = BuildingService(), locationService: LocationService = LocationService()) {
```
[NIT] You can remove the type since Swift can infer the type from the initialised units

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingInteractor.swift:
```diff
@@ -5,18 +5,67 @@
+      return .failure(error)
+    }
+
+    // // More efficient?
+    // let buildingsWithDistances = buildings.map { building -> (building: Building, distance: CLLocationDistance) in
+    //    let distance = calculateDistance(from: currentLocation, to: building)
+    //    return (building, distance)
+    // }
+    //
+    // let sortedBuildingsWithDistances = buildingsWithDistances.sorted { lhs, rhs in
+    //    return inAscendingOrder ? lhs.distance < rhs.distance : lhs.distance > rhs.distance
+    // }
+    //
+    // return sortedBuildingsWithDistances.map { $0.building }
```
This is indeed more efficient since we save on repeated calculateDistance calls, you can uncomment this implementation and remove the existing one.

avdn-dev reviewed (changes requested):
Good thinking with the more efficient sorting implementation! Just need to figure out a way to abstract away the CoreLocation distance calculation.

mmc03-ucb commented:
- Removed `CoreLocation` import and `CLLocation` usage  
- Added pure-Swift `calculateDistance` helper returning `Double`  
- Updated init to use type inference (`.init()`)  
- Switched to one-pass map→sort→map distance sorting

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingInteractor.swift:
```diff
@@ -6,17 +6,57 @@
+  private func calculateDistance(from location: Location, to building: Building) -> Double {
+    let dlat = building.latitude  - location.latitude
+    let dlon = building.longitude - location.longitude
+    return sqrt(dlat * dlat + dlon * dlon)
```
[NIT] Since this distance representation is only used within the BuildingInteractor, you can omit the sqrt for even more performance (on n < 100 lol)
